contracts/src/registry/L1ETHRegistry.sol

```diff
@@ -0,0 +1,244 @@
+    {
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
+        }
+        return super.ownerOf(tokenId);
+    }
+
+    /**
+     * @dev Receive an ejected name from L2.
+     * @param label The label of the name
+     * @param owner The owner of the name
+     * @param registry The registry to use for the name
+     * @param flags Additional flags for the name
+     * @param expires Expiration timestamp
+     * @return tokenId The token ID of the ejected name
+     */
+    function ejectFromL2(string calldata label, address owner, IRegistry registry, uint96 flags, uint64 expires)
+        public
+        onlyRole(EJECTION_CONTROLLER_ROLE)
+        returns (uint256 tokenId)
+    {
+        tokenId = (uint256(keccak256(bytes(label))) & ~uint256(FLAGS_MASK)) | flags;
+        flags = (flags & FLAGS_MASK) | (uint96(expires) << 32);
+
+        // If there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        datastore.setSubregistry(tokenId, address(registry), flags);
+        emit NewSubname(label);
+        emit NameEjected(tokenId, owner, expires);
+        return tokenId;
+    }
+
+    /**
+     * @dev Renew an ejected name
+     * @param tokenId The token ID of the name to renew
+     * @param expires New expiration timestamp
+     */
+    function renew(uint256 tokenId, uint64 expires) public onlyRole(RENEWAL_CONTROLLER_ROLE) {
+        (address subregistry, uint96 flags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiration = _extractExpiry(flags);
+        if (oldExpiration < block.timestamp) {
+            revert NameExpired(tokenId);
+        }
+        if (expires < oldExpiration) {
+            revert CannotReduceExpiration(oldExpiration, expires);
+        }
+        datastore.setSubregistry(tokenId, subregistry, (flags & FLAGS_MASK) | (uint96(expires) << 32));
+
+        address observer = tokenObservers[tokenId];
+        if (observer != address(0)) {
+            L1ETHRegistryTokenObserver(observer).onRenew(tokenId, expires, msg.sender);
+        }
+
+        emit NameRenewed(tokenId, expires, msg.sender);
+    }
+
+    /**
+     * @dev Set the token observer for a name
+     * @param tokenId The token ID of the name
+     * @param _observer The observer address
+     */
+    function setTokenObserver(uint256 tokenId, address _observer) external onlyTokenOwner(tokenId) {
+        tokenObservers[tokenId] = _observer;
+        emit TokenObserverSet(tokenId, _observer);
+    }
+
+    /**
+     * @dev Relinquish an ejected name
+     * @param tokenId The token ID of the name to relinquish
+     */
+    function relinquish(uint256 tokenId) external onlyTokenOwner(tokenId) {
```

Comment:
How does this differ from `migrateToL2`?

Reply:
Why would we want to let people do this? It would seem it puts the name in a weird unrecoverable no-mans-land.

Reply:
the idea was to give the flexibility of what L2ETH does can be done on ejected names too but, went through the migration design again, seemingly this complicates the coordination between L1 and namechain. user should migrate back to L2 for relinquish. removing it.